src/handlers/VehicleTestController.ts

```diff
@@ -1,58 +1,230 @@
+  public getExpiryStrategy(
+    testTestForExpiry: TestTypeForExpiry
+  ): IExpiryDateStrategy {
+    const selectedStrategy = ExpiryDateStrategyFactory.GetExpiryStrategy(
+      testTestForExpiry,
+      this.dateProvider
+    );
     return selectedStrategy;
   }
+
+  public async insertTestResult(payload: models.ITestResultPayload) {
+    try {
+      utils.ValidationUtil.validateInsertTestResultPayload(payload);
+      console.info("validation is success");
+      payload = utils.MappingUtil.setCreatedAtAndLastUpdatedAtDates(payload);
+      const testTypeParams: models.TestTypeParams = {
+        vehicleType: payload.vehicleType,
+        vehicleSize: payload.vehicleSize,
+        vehicleConfiguration: payload.vehicleConfiguration,
+        vehicleAxles: payload.noOfAxles,
+        euVehicleCategory: payload.euVehicleCategory,
+        vehicleClass: payload.vehicleClass?.code,
+        vehicleSubclass: payload.vehicleSubclass?.[0],
+        vehicleWheels: payload.numberOfWheelsDriven,
+      };
+      const testTypesWithTestCodesAndClassification = await this.dataProvider.getTestTypesWithTestCodesAndClassification(
+        payload.testTypes,
+        testTypeParams
+      );
+      payload.testTypes = testTypesWithTestCodesAndClassification;
+      const payloadWithTestNumber = await this.dataProvider.setTestNumberForEachTestType(
+        payload
+      );
+      const payloadWithExpiryDate = await this.generateExpiryDate(
+        payloadWithTestNumber
+      );
+      const payloadWithCertificateNumber = VehicleTestController.AssignCertificateNumberToTestTypes(
+        payloadWithExpiryDate
+      );
+      const payloadWithAnniversaryDate = VehicleTestController.calculateAnniversaryDate(
+        payloadWithCertificateNumber
+      );
+      payloadWithAnniversaryDate.vehicleId = payloadWithAnniversaryDate.vrm;
+      const result = await this.dataProvider.insertTestResult(payloadWithAnniversaryDate);
+      return result;
+    } catch (error) {
+      if (error.statusCode === 400 && error.message === enums.MESSAGES.CONDITIONAL_REQUEST_FAILED) {
+          console.info(
+            "TestResultService.insertTestResult: Test Result id already exists",
+            error
+          );
+          error =  new models.HTTPResponse(201, enums.MESSAGES.ID_ALREADY_EXISTS);
+        }
+      throw error;
+    }
+  }
+//#endregion
+//#region [rgba(0, 205, 30, 0.1)] Private functions
+  /**
+   * Note: When performing actions on a moment instance, it gets mutated
+   * Note: Expiry dates on the payload should be set at the UTC start of day.
+   *
+   * @param payload
+   */
+  private async generateExpiryDate(
+    payload: models.ITestResultPayload
+  ): Promise<models.ITestResultPayload> {
+    try {
+      if (
+        payload.testStatus !== enums.TEST_STATUS.SUBMITTED ||
+        utils.ValidationUtil.isNotAllowedVehicleTypeForExpiry(
+          payload.vehicleType
+        )
+      ) {
+        return Promise.resolve(payload);
```

Comment:
I think we don't need the Promise.resolve, by default async functions return promises

Reply:
removed